opensearch-operator/controllers/clusterController.go

```diff
@@ -4,122 +4,78 @@ import (
-		if err != nil {
-			if !errors.IsAlreadyExists(err) {
-				fmt.Println(err, "Cannot create Headless Service")
-				r.Recorder.Event(r.Instance, "Warning", "Cannot create Headless Service ", "Requeue - Fix the problem you have on main Opensearch Headless Service ")
-
-				return ctrl.Result{}, err
-			}
-		}
-		fmt.Println("service Created successfully", "name", headless_service.Name)
-	}
-
 	service := v1.Service{}
-	serviceName = r.Instance.Spec.General.ServiceName + "-svc"
+	serviceName := r.Instance.Spec.General.ServiceName
 	if err := r.Get(context.TODO(), client.ObjectKey{Name: serviceName, Namespace: namespace}, &service); err != nil {
-
-		/// ------ Create External Service -------
+		// Create External Service
 		clusterService := builders.NewServiceForCR(r.Instance)
 
 		err = r.Create(context.TODO(), clusterService)
 		if err != nil {
 			if !errors.IsAlreadyExists(err) {
-				fmt.Println(err, "Cannot create service")
+				r.Logger.Error(err, "Cannot create service")
 				r.Recorder.Event(r.Instance, "Warning", "Cannot create opensearch Service ", "Requeue - Fix the problem you have on main Opensearc Service ")
-				return ctrl.Result{}, err
+				return nil, err
 			}
 
 		}
-		fmt.Println("service Created successfully", "name", service.Name)
-
+		r.Logger.Info(fmt.Sprintf("service %s created successfully", clusterService.Name))
 	}
 
-	///// ------ Create Es Nodes StatefulSet -------
-	NodesCount := len(r.Instance.Spec.NodePools)
-	sts := sts.StatefulSet{}
+	// Create StatefulSets for NodePools
+	for _, nodePool := range r.Instance.Spec.NodePools {
+		// Create headless service for sts
+		targetService := builders.NewHeadlessServiceForNodePool(r.Instance, &nodePool)
+		existingService := v1.Service{}
+		if err := r.Get(context.TODO(), client.ObjectKey{Name: targetService.Name, Namespace: namespace}, &existingService); err != nil {
+			err = r.Create(context.TODO(), targetService)
+			if err != nil {
+				if !errors.IsAlreadyExists(err) {
+					r.Logger.Error(err, fmt.Sprintf("Cannot create Headless Service for nodepool %s", nodePool.Component))
+					r.Recorder.Event(r.Instance, "Warning", "Cannot create Headless Service ", "Requeue - Fix the problem you have on main Opensearch Headless Service ")
+					return nil, err
+				}
+			}
+			r.Logger.Info(fmt.Sprintf("service for nodepool %s created successfully", nodePool.Component))
+		}
 
-	for x := 0; x < NodesCount; x++ {
-		sts_for_build := builders.NewSTSForCR(r.Instance, r.Instance.Spec.NodePools[x])
-		stsName := r.Instance.Spec.General.ClusterName + "-" + r.Instance.Spec.NodePools[x].Component
-		if err := r.Get(context.TODO(), client.ObjectKey{Name: stsName, Namespace: namespace}, &sts); err != nil {
-			/// ------ Create Es StatefulSet -------
-			fmt.Println("Starting create ", r.Instance.Spec.NodePools[x].Component, " Sts")
-			err = r.Create(context.TODO(), sts_for_build)
+		stsName := r.Instance.Spec.General.ClusterName + "-" + nodePool.Component
```

Comment:
a small explanation about the "targetSTS" and "exisitingSTS" will be great :)

Reply:
This is anticipating a bit that later on the cluster controller will need to make updates to the SatefulSet when something with the cluster spec changes. So the code first builds the STS spec as it should be based on the current spec (`targetSTS`) then checks if the STS already exists in the cluster (`existingSTS`). In the future the controller could then compare both to see if anything needs to change.